### PR TITLE
Add Ed25519 support

### PR DESCRIPTION
**Pull request by mildred**
This pull request add [ed25519](http://ed25519.cr.yp.to/) support to IPNS. This probably needs some polishing. In particular, I don't understand how to implement the Encode and Decode functions (see code). I don't even see why they are useful in our context. We could leave them out completely, no one uses them.

### CODE REVIEW

File: p2p/crypto/ed25519.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+func (pk *Ed25519PublicKey) Verify(data, sig []byte) (bool, error) {
+	if len(sig) != ed.SignatureSize {
+		return false, errors.New("Incorrect signature size")
+	}
+	var sig2 ed.Signature
+	copy(sig, sig2[:])
+	return ed.Verify(sig2, data, pk.k), nil
+}
+
+func (pk *Ed25519PublicKey) Bytes() ([]byte, error) {
```
Comment (whyrusleeping) [inline p2p/crypto/ed25519.go]: For the bytes method, its not well documented, but we want to use the protobuf structure in `p2p/crypto/internal/pb`. This allows us to easily detect and decode private and public keys (self describing data!)

Comment (whyrusleeping) [inline p2p/crypto/ed25519.go]: Whoops, i was looking at the wrong code. This is fine.

**Comment by whyrusleeping** [comment]
Could you also add a test or two for ed25519? In p2p/crypto/key_test.go i have a function called `TestRsaKeys` which is just a driver for other test functions which accept an interface keytype. Just pass an ed25519 key to those

**Comment by mildred** [comment]
About Encrypt and Decrypt and how it is done in Ed25519. This is all based on a key_exchange function. With two key pairs K1{SK1,PK1} and K2{SK2,PK2} we can exchange a shared secret that can be used to encrypt or decrypt:

    key_exchange(PK1,SK2) == secret == key_exchange(PK2,SK1)

I wrongly assumed that we needed our own secret key SK1 to encrypt to a public key PK2. But that is not the case, as we can create a key for the purpose of the encryption.

Bob wants to encrypt message to Alice. He has the public Alice key PK2:

    SK1, PK1 = create_keypair()
    secret = key_exchange(PK2, SK1)
    encrypted_message = [PK1, encrypt(secret, message)]
    // SK1 is secret and can be used to decrypt the message

Alice wants to decrypt the message. She knows her own key SK2, PK2 and can deduce PK1 from the encrypted message:

    PK1 = encrypted_message[0]
    secret = key_exchange(PK1, SK2)
    message = decrypt(secret, encrypted[message[1])

The functions `encrypt(secret, msg)` and `decrypt(secret, msg)` are symmetric encryption function that take a shared secret.

**Comment by whyrusleeping** [comment]
Interesting. Its probably best to separate out the Encryption functionality from the Key interface at this point.

**Comment by mildred** [comment]
Ok, I'll update the pull request and refactor this interface

**Comment by jbenet** [comment]
@mildred  i'm wary of adding crypto implementations from a lib not in high use. [1] Without sinking into it myself and vetting it, I've no clear indication that orlp/ed25519 is to be trusted. (cc @orlp so he can come discuss if he wishes) I wonder if the right thing to do here is to just get to `multikey`[2] faster (the latest installment of the `multi--` formats) and allow pluggable key implementations, so you could build a version of `go-ipfs` with these keys trivially, but doesnt require putting it in the main implementation just yet.[3] 


[1] in fact i'm considering removing our `secio` pipe altogether in favor of https://godoc.org/golang.org/x/crypto/ssh (and thus just use ssh keys inside a `~/.go-ipfs/keys` directory.

[2] basically, ensure go-ipfs allows configurable/pluggable algorithms + upgradeability.

[3] I will regard all the crypto in go-ipfs as suspect and not to be trusted until I convince myself otherwise through multiple independent careful audits.

**Comment by mildred** [comment]
I understand that you are wary of crypto, and rightly so. Using well-known code is better and your suggestion of using the openssh lib seem to be a good one. This shouldn't prevent implementing ed25519 implementation as it is included in [OpenSSH](http://article.gmane.org/gmane.network.openssh.announce/71/match=ed25519) itself (but perhaps not in the go package).

For the moment, this pull request add support for Ed25519 as a way to recognize keys in IPNS. This doesn't involve the private key generated and stored in ~/.go-ipfs. As such it is relatively harmless to include it now.

What do you suggest for this pull request?

**Comment by mildred** [comment]
The issue that keep track of missing ed25519 algorithm in the go `crypto/ssh` package: https://github.com/golang/go/issues/8581

**Comment by orlp** [comment]
@jbenet I am summoned.

My Ed25519 code does two things:

1. A simple API wrapper around djb's original Ed25519 code, which works fine but has a terrible interface. djb is a good cryptographer, but doesn't use the best programming practices.

2. Two extra features, Curve25519 key exchange using Ed25519 keys, and adding a scalar to a keypair. This functionality is written by me, however using the field functions written by djb. They are likely to be correct, but are not written by a world-renowned cryptographer like djb.

Ed25519 is not a key exchange, it is a signing algorithm. Curve25519 is a key exchange, but can not sign. However, Curve25519 and Ed25519 are strongly related, and Ed25519 public/private keys can be independently converted to a similar Curve25519 keypair. My key exchange functions do this automatically and then use Curve25519 to do the key exchange.

Adding a scalar to a keypair is an advanced feature that you should not use unless you know what it is that you need it. I mostly wrote it for my personal use, which is one party assuring a keypair is randomly generated without knowing the private key.

Now, should you use my library? I haven't looked at your code, but the answer depends on your usage:

1. __You only use Ed25519 signing and verification.__ You can go ahead and use my library. All that functionality is implemented by djb and is correct with near certainty that my API changes do not violate that.

2. __You only use Curve25519 key exchange.__ I would recommend using a dedicated Curve25519 library, like djb's reference implementation: http://cr.yp.to/ecdh.html

3. __You use both Ed25519 signing and verification and Curve25519 key exchange.__ My library does this in a simple and clean way. My implementation of the conversion and the signing is almost guaranteed to be correct.

    A couple words of warning, though. Signing and key exchanging with the same keypair is something that traditionally is not done, because one algorithm might leak information about the other's private key. Ed25519/Curve25519 seems to be an exception to this, but I'm not aware of a proof of security by a renowned cryptographer. See [here](https://moderncrypto.org/mail-archive/curves/2014/000205.html) and [here](http://crypto.stackexchange.com/questions/3260/using-same-keypair-for-diffie-hellman-and-signing) for more information. Unless you guys absolutely want using the same keypair for both signing and key exchanging, using different keypairs for both is probably safest.

__And finally, the thoughest question, should I and my library be trusted for actual security-sensitive codebases?__ The answer to that is dependent on how paranoid you are. I can promise you that I didn't maliciously enter backdoors into the code. However, you have to ask yourself what my promise is worth to your users if their privacy was violated. And that just covers malice. I consider myself a good programmer, but everyone makes mistakes. After all I'm just another person on the internet.

As much as it pains me to say so, if security is an absolute must my suggestion would be to sacrifice my tasty beautiful API and trade it in for a more publicly scrutinized library with a more ugly interface, like https://github.com/jedisct1/libsodium .

**Comment by whyrusleeping** [comment]
@orlp, thank you for the response! Our usage of ed25519 is going to be for signing and verification, just making sure people are talking to who they think they are. The interface for our  keys currently contains an Encryption method, but its not used and will soon be removed.
My main concern with using ed25519 currently, is that its written in C, and our go code will lose its wonderful portability (which is nothing against C or ed25519).

**Comment by orlp** [comment]
Adam Langley has an Ed25519 library written in Go: https://github.com/agl/ed25519 .

**Comment by orlp** [comment]
I looked a bit more at the project, and it seems you guys are doing things similar to my (currently not-that-active) aspirations. So I'd like to share one of my ideas which sprouted my library and the `add_scalar` function.

My main problem with DHTs are the sybil attacks and eclipse attacks. One of my solutions involves a centralized identity provider (or multiple, like OpenID) that signs your public key to prove you may connect to the network (your certificate), and you sign your messages with your public key.

In order to save bandwidth overhead to the max I made Kademlia ID == public key. This is dangerous thanks to the eclipse attack. In order to mitigate this I used the `add_scalar` function I added to my library. This function allows the identity provider to randomize the public key of a keypair without knowing the private key by adding a random scalar to it, signing the resulting public key, and giving back this certificate + the random scalar so the user can figure out his private key. This solves the eclipse attack.

**Comment by whyrusleeping** [comment]
@orlp Using a centralized identity provider is something discussed in the S/Kademlia paper, which we are basing our DHT security off of (not entirely implemented yet, but were getting there) my issue with that is we dont want to rely on any single point of failure, if the identity provider goes down, the network starts to die. 

If large scale distributed systems interest you, we would love to have you around! If you want to chat more, we have an irc channel on freenode at #ipfs

**Comment by mildred** [comment]
@whyrusleeping About portability because of use of C code, I think what you are referring to is usage of dynamic linking with external libraries. Two things on that:

- Are you sure that ipfs isn't already using a dynamically linked libc (`ldd .../bin/ipfs` will tell you). I heard without being certain that some parts of the standard Go library used C code around the http packages.

- The Ed25519 implementation doesn't need any function from the libc as far as I know. This means that linking to the libc is not at all necessary. If compiling ed25519 code using cgo leads to a dynamic link with a libc, this should be treated as an issue with the cgo compiler I believe.

**Comment by orlp** [comment]
My Ed25519 library does not use libc, with the exception of `ed25519_create_seed`. If you have already have a cryptographically secure random number generator you do not need this function. You can define `ED25519_NO_SEED` to prevent this function from being generated, and this removes all possible dependency on libc or other OS libraries.

**Comment by mildred** [comment]
I was thinking on how we could make sure to use the safest possible implementation of ed25519. Obviously, if you use ed25519, you trust djb, the author of the algorithm, with it. Code written by djb should then be safe to include in out codebase.

This leaves with code referenced at: http://ed25519.cr.yp.to/software.html

Te code is in the SUPERCOP benchmark tools at [crypto_sign/ed25519](https://github.com/floodyberry/supercop/tree/master/crypto_sign/ed25519). This link is for looking up code online. You should download [the reference SUPERCOP source instead](http://bench.cr.yp.to/supercop.html).

Quoting djb:

> The Ed25519 software consists of three separate implementations, all providing the same interface:
> * amd64-51-30k. Assembly-language implementation for the amd64 architecture, using radix 2^51 and a 30KB precomputed table.
> * amd64-64-24k. Assembly-language implementation for the amd64 architecture, using radix 2^64 and a 24KB precomputed table.
> * ref. Slow but relatively simple and portable C implementation. 

I suggest we take the code in ref (the algorithm is already fast, I don't think we should bother with the optimized versions for now), and write a cgo binding to it. As the code doesn't require any C library functions (except to generate a random seed, but we can do that in Go code), we should be able to generate a go executable that doesn't depend on the libc.

@whyrusleeping what do you think?

Mildred

**Comment by whyrusleeping** [comment]
Yeah, if we can avoid linking in libc, im happy. Im not a huge stickler on explicitly using the reference code, but since I use an arm laptop, im not gonna be super excited about an amd64 architecture only implementation, lol. Lets ask @jbenet what he thinks of binding the C code vs using a lib written entirely in go.
Im going to read up on exactly how the algorithm works so I can make more informed decisions about integrating it into our system.

**Comment by orlp** [comment]
@mildred If you're going to directly use djb's code (mind you, my library is exactly that, only reformatted with a different API, but you'd need to audit it to make sure), use the ref10 implementation, it's a __lot__ faster than the ref implementation, and pure C.

**Comment by whyrusleeping** [comment]
I would feel comfortable using @orlp's library, as long as we can swap out the random seed stuff with go code to avoid linking libc

**Comment by jbenet** [comment]
@whyrusleeping is avoiding clib just to avoid linking clib in? afaiu we'd have to compile with cgo anyway, right?

@mildred: i'd be fine with the approach you mention or with using @orlp's code directly. We have to do a full audit of all the crypto in ipfs anyway. i think this should happen after we: 
- (a) move the keys out of the config file into a `keys/` directory in the repo.
- (b) make sure we use existing key serialization standards everywhere (i think we roll our own in a couple places, so we end up displaying base64 encoded things to the user that aren't clearly keys), and 
- (c) make the key algorithms selectable (right now we default to rsa everywhere). 

It would be good to discuss all the things that need to happen for (a), (b), (c) before starting on any.

@mildred @orlp aside, thoughts on https://github.com/jedisct1/libsodium ? (i havent looked in depth into it at all, was recently mentioned to me)

**Comment by whyrusleeping** [comment]
@jbenet, my thoughts on libc are that if we have to link it in, we lose a degree of portability. If we use cgo but manage to avoid needing libc, we can still retain our "cross platform"  binary

**Comment by jbenet** [comment]
> @jbenet, my thoughts on libc are that if we have to link it in, we lose a degree of portability. If we use cgo but manage to avoid needing libc, we can still retain our "cross platform" binary

Ah right.

**Comment by mildred** [comment]
About libsodium, it's a fork of NaCl library by Daniel J. Bernstein. It seems improved in some way, and it contains an ed25519 implementation as well. The upstream NaCl library doesn't yet as far as I know. Probably because no release has been made for some time.

The ed25519 implementation is probably the same, but the code base is larger because the library can do much more than just sign using ed25519 keys. As it is larger, we either have to link to it all (with many functions that would go unused, and probably some libc calls we won't be able to avoid) or restrict to a few files that seem identical to what @orlp has.

**Comment by jbenet** [comment]
Yeah libsodium would be only in the case we need more constructions, which we may.

**Comment by mildred** [comment]
I was wrong somewhere. On IRC, someone told me some time ago that if the C code didn't use libc calls, it wouldn't link against the libc and I believed it when I shouldn't have. I'm now trying to have a simple Go command to test how to link to C code without linking to the libc. [I asked on golang-nuts for help.](https://groups.google.com/forum/#!topic/golang-nuts/_qQH9rnM_zQ)

Also, orlp told us of an ed25519 implementation written natively in Go: https://github.com/agl/ed25519 I don't know if we could trust this code as much as the code written by Daniel J. Bernstein.

**Comment by jbenet** [comment]
@mildred agl is a crypto guy at google. If you're using chrome or google search you probably use his crypto. He also wrote: https://godoc.org/golang.org/x/crypto/curve25519 but it does not include the ops (i wonder where it's used).

I've asked him about the status of things here: https://groups.google.com/d/msg/golang-nuts/eN7d_3Oh9c4/FtmcnigRSWAJ

**Comment by orlp** [comment]
Whatever you do, please do not use a C-to-Go transpiler. That is a security disaster waiting to happen.

I think you guys have four options:

1. Use my library. Requires security audit, but easy to compare to djb's ref10 implementation. Requires C.

2. Use djb's ref10 implementation. Does not require security audit. Requires C. Ugly interface and code structure.

3. Use Adam Langley's https://github.com/agl/ed25519 . Requires security audit.

4. Use https://github.com/jedisct1/libsodium . Potential code bloat. Ugly interface.

**Comment by whyrusleeping** [comment]
I like the idea of using @agl's pure go library. If it in fact needs a security audit, thats alright, we do too.

**Comment by mildred** [comment]
This is continued in #827

**Pull request closed without merging**